### DISCUSSION TITLE
GPUAdapter.limits seems unable to detect the true capabilities of the GPU

### DISCUSSION THREAD

**Discussion by microaaron** [gpuweb/gpuweb#4533] [Q&A]
![捕获GPUlimit](https://github.com/gpuweb/gpuweb/assets/16946786/b7c81165-95e0-451e-9b61-ada70c13479b)
The GPUSupportedLimits obtained by Chrome(dawn) and Firefox(wgpu) are not consistent.
And the capabilities of modern GPUs are obviously higher than these data.
This data appears to be fixed in the software rather than being read from the hardware. And it's set very conservatively.(e.g. maxUniformBuffersPerShaderStage, [maxVertexBuffers](https://github.com/gfx-rs/wgpu/issues/558)...)
This makes it difficult to convert some webgl programs into webgpu programs

**Comment by Kangz** [comment]
This is completely expected. Browser can expose any limits lower than what the hardware supports, and for a variety of reasons. They are required to bucket limits and features to limit the amount of fingerprinting exposed, they could have implementation limits (for example `maxBindGroups:4` in Chromium at the moment), and any other reason.

It would be useful to know what limit is making it difficult to port your WebGL code so we could suggest a workaround, or add some new buckets in implementations.

**Reply by microaaron** [reply]
I encountered the problem that Chrome's ```maxUniformBuffersPerShaderStage``` and ```maxVertexBuffers``` are too small when porting WebGL codes.
I think webgpu should somehow get the full capabilities of the hardware to maximize the performance of the GPU.
As for inferring fingerprinting based on GPU capabilities, I think it is meaningless, because Webgl's API can completely obtain detailed GPU information.

**Reply by Kangz** [reply]
What values do you get in Chromium for these in WebGL / WebGPU and what values do you need?

**Reply by microaaron** [reply]
For ```maxUniformBuffersPerShaderStage``` and ```maxVertexBuffers```, I hope Chromium provides at least the same values as firefox. 
Firefox provides 64 UniformBuffersPerShaderStages and 16 VertexBuffers. Chromium only provides 12 and 8.
Of course, this may still not meet future needs

**Comment by greggman** [comment]
It's been brought up before for `maxVertexBuffers` to make the default be 16 to match WebGL2

https://github.com/gpuweb/gpuweb/issues/4284

`maxUniformBuffersPerStage` it doesn't seem like we could raise the default but certainly various browsers could look into raising the limit when supported.

**Reply by microaaron** [reply]
> `maxUniformBuffersPerStage` it doesn't seem like we could raise the default but certainly various browsers could look into raising the limit when supported.

I don't expect to raise the default value. The default value should be set conservatively to be compatible with older hardware. But the browser should be able to detect the maximum capabilities of the GPU. Currently Chromium can only support default values, and Firefox can pass ```requiredLimits``` in ```requestDevice()``` to support larger values.

**Comment by arjami25** [comment]
great!
